[chin-herng] iP

chin-herng opened the pull request:
# A first-level heading
> Text that is a quote. This site was built using [GitHub Pages](https://pages.github.com/).
- George Washington
* John Adams
+ ~~This was mistaken text~~ _This text is italicized_
1. James Madison
2. James Monroe
3. John Quincy Adams :+1:
- [x] #739
- [ ] https://github.com/octo-org/octo-repo/issues/740
- [ ] Add delight to the experience when all tasks are complete :tada:

Use `git status` to list all new or modified files that haven't yet been committed.

```ruby
require 'redcarpet'
markdown = Redcarpet.new("Hello World!")
puts markdown.to_html
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

chin-herng committed:
Rename chatbot, implement greet and exit

chin-herng committed:
Implement printFormatted

chin-herng committed:
Overload printFormatted to accept a single String

chin-herng committed:
Prepare test input and expected output

chin-herng committed:
Configure text UI testing

chin-herng committed:
Update expected output

chin-herng committed:
Implement echo

chin-herng committed:
Customise chatbot

chin-herng committed:
Update expected output

chin-herng committed:
Implement add and list

chin-herng committed:
Change how scanner and tasks are initialised

chin-herng committed:
Remove unnecessary toString for formatted

chin-herng committed:
Customise chatbot

chin-herng committed:
Add check for exceeding maximum number of tasks

chin-herng committed:
Update test

chin-herng committed:
Create Commands and Replies classes

chin-herng committed:
Implement Commands and Replies class for more OOP

chin-herng committed:
Implement mark as done/not done

chin-herng committed:
Implement indent and line break utilities

chin-herng committed:
Implement constants for done and not done icons

chin-herng committed:
Customise chatbot

chin-herng committed:
Add list header

chin-herng committed:
Implement ToDos, Events, Deadlines

chin-herng committed:
Implement utility function to extract parameters

chin-herng committed:
Rename parameter of extractParameters

chin-herng committed:
Combine logic for marking as done or not done

chin-herng committed:
Edit replies for add

chin-herng committed:
Change variable name in list

chin-herng committed:
Implement minimal error handling

chin-herng committed:
Add testing for empty description of deadline and event

chin-herng committed:
Move command processing logic to Commands class

chin-herng committed:
Implement typical error handling

chin-herng committed:
Fix InvalidTaskIndexException message

chin-herng committed:
Make InvalidTaskIndexException message more specific

chin-herng committed:
Change PARAMETER_NOT_FOUND to use format specifier

chin-herng committed:
Change TASKS to be ArrayList<Task>

chin-herng committed:
Implement delete

chin-herng committed:
Implement save

chin-herng committed:
Merge branch 'branch-Level-7'

chin-herng committed:
Update tests

chin-herng committed:
Update tests

chin-herng committed:
Use Dates and Times instead of Strings

chin-herng committed:
Merge branch 'branch-Level-8'

Resolve merge conflict

chin-herng committed:
Remove unnecessary overloaded Deadline constructor

chin-herng committed:
Rename Replies to Ui

chin-herng committed:
Extract Storage class

chin-herng committed:
Rename Commands to Parser

chin-herng committed:
Extract TaskList class

chin-herng committed:
Add TODO to separate methods for adding tasks

chin-herng committed:
Merge branch 'branch-A-MoreOOP'

chin-herng committed:
Put all classes in a package bob

chin-herng committed:
Merge branch 'branch-A-Packages'

chin-herng committed:
Merge branch 'add-gradle-support'

chin-herng committed:
Set up gradle

miffi commented on src/main/java/bob/Bob.java:
```diff
@@ -0,0 +1,38 @@
+import java.time.format.DateTimeParseException;
+
+public class Bob {
+    public static void handleMark(int taskIndex, boolean isDone) throws InvalidTaskIndexException {
```
I feel that the `handle` prefix is redundant. `mark` is a verb already, and it describes what is being done. Therefore, maybe returning mark will make it more concise. If the usage of the function is to be made clearer,  I think maybe something like `markTask` would be better. (Then you could also change taskIndex to just index as it would be pretty obvious what it's indexing imo)

Similar sentiment for the other functions this class.

chin-herng replied:
hi! im sort of establishing a convention that whatever commands has to go through the Bob class first, and Bob will coordinate everything - he "handles" the command over to other classes. for example Bob "handles" the mark command by telling TaskList to actually do the marking

hope it makes sense haha, that's my justification for why every method in Bob has a handle prefix, it's to emphasise that the method is a command handler.

miffi commented on src/main/java/bob/Parser.java:
```diff
@@ -0,0 +1,134 @@
+    public static final String EVENT = "event";
+    public static final String DELETE = "delete";
+
+    public static final DateTimeFormatter INPUT_DATE_FORMATTER = DateTimeFormatter.ISO_LOCAL_DATE;
```
I've seen this constant being used for both input and output, so maybe just `DATE_FORMATTER` might be better.

chin-herng replied:
hi! the input and output date format are different. the tasks read from the command line, and are stored in the hard disk in the format of input date format. the tasks are being printed in the format of output date format.

miffi commented on src/main/java/bob/Event.java:
```diff
@@ -0,0 +1,26 @@
+    }
+
+    public String format() {
+        return "event | " + super.format() + " | " + this.from.format(Parser.INPUT_DATE_FORMATTER)
```
Okay this line has thrown me in a bit of a loop. You have used both INPUT and OUTPUT for output in this case so I'm not really sure what different significance they have. Maybe you can try changing the names or splitting into other variables to make this more obvious?

chin-herng replied:
yea youre right that the naming can be better. the format method is to format the task in the format stored in the hard disk. input format is used so that when loading the tasks from the hard disk, we can load the tasks as if we are reading the tasks in the file.

miffi commented on src/main/java/bob/Ui.java:
```diff
@@ -0,0 +1,89 @@
+    public static final String DELETE_FOOTER = "%d task(s) left";
+
+    public static final String EMPTY_DESCRIPTION = "missing description for %s";
+    public static final String INVALID_COMMAND = "what";
```
Honestly having all of the `INVALID`s in the UI is an interesting choice. It brings all of the CLI text into one place which would make it easier to edit and understand how the text looks as a whole. However, it does make understanding what the exceptions say when reading the code. I don't have a particular recommendation, but imo having the messages where they're used, i.e. the exceptions, would be better.

chin-herng replied:
thanks! i considered putting these messages in the corresponding exceptions but i was thinking that it's more convenient to customise my bot if i put all the ui-related outputs in the Ui class. there might be alternative ways to do this and i might do the refactor in the future.

miffi commented on src/main/java/bob/Deadline.java:
```diff
@@ -0,0 +1,23 @@
+        this.by = LocalDate.parse(by, Parser.INPUT_DATE_FORMATTER);
+    }
+
+    public String format() {
```
I see that this is overriding the superclass. Maybe you could add @Override to this and to the same function in other subclasses to make this obvious.

miffi commented on src/main/java/bob/Parser.java:
```diff
@@ -0,0 +1,134 @@
+
+    public static final DateTimeFormatter INPUT_DATE_FORMATTER = DateTimeFormatter.ISO_LOCAL_DATE;
+
+    public static String[] extractParameters(String parametersString,
```
It's a bit hard to understand what the parameters do. From the code it seems that the `parametersString` is the input to extract the parameters from, but I don't see any use for parameters other than its length. Adding a javaDoc header might help. If the only use for parameters is the length I think that passing the length itself is better.

chin-herng replied:
yea, this method deserves a javadoc, will add it in the future!

miffi commented on src/main/java/bob/Parser.java:
```diff
@@ -0,0 +1,134 @@
+        return result;
+    }
+
+    public static void processDeleteOrMarkCommands(String[] commandArgs) throws NumberFormatException,
```
`Process` is vague. It doesn't really indicate what the `processing` is. I think changing to `make` or maybe `create` would describe what this is doing better.

chin-herng replied:
youre right that the word process is chosen due to lack of vocabulary. on second thought it might be better to use "parse".

miffi commented on src/main/java/bob/Parser.java:
```diff
@@ -0,0 +1,134 @@
+        }
+    }
+
+    public static boolean listen() {
```
I think this name is fine. It describes what the command is doing. However, it is not obvious what it is supposed to do from just its name. A javaDoc header imo should help with that.

miffi commented on src/main/java/bob/Parser.java:
```diff
@@ -0,0 +1,134 @@
+                processDeleteOrMarkCommands(commandArgs);
+            } catch (NumberFormatException e) {
+                // The more "correct" way is to throw an InvalidTaskIndexException?
+                Ui.print(String.format(Ui.INVALID_TASK_INDEX, commandArgs[1]));
```
Instead of catching the `NumberFormatException` and printing out the index here, I feel that it might be better to have another exception whose message itself has the index with the specific command. That would would express the intent of the exception in the code better imo.

chin-herng replied:
thanks for the suggestion!

miffi reviewed (commented):
Hello :). I'm a peer reviewer for your iP! I'm doing my second review for you (the code quality one). I have some recommendations, but overall I think that your code is readable and well structured.

One overarching suggestion is that you could move Task and its subclasses into a bob.task package. I feel like that would make their relation more obvious.

chin-herng committed:
Fix load() triggers save()

chin-herng committed:
Rename process to parse

chin-herng committed:
Remove Commands from Parser method names

chin-herng committed:
Refactor on adding tasks

chin-herng committed:
Implement list on date

chin-herng committed:
Implement list due in

chin-herng committed:
Refactor code to be more OOP

chin-herng committed:
Remove unnecessary import

chin-herng committed:
Extract out logic for runCommandLoopUntilExitCommand

chin-herng committed:
Implement trimming on user input

chin-herng committed:
Add TODO comment to use regex in Parser

chin-herng committed:
Add comment to indicate possibility of CommandResult classes

chin-herng committed:
Divide into multiple packages

chin-herng committed:
Test behaviour of toStorageFormat

chin-herng committed:
Merge branch 'branch-A-JUnit'

chin-herng committed:
Test isOccurringOn

chin-herng committed:
Test isDueIn

chin-herng committed:
Test toString

chin-herng committed:
Test parse

chin-herng committed:
Test Storage

chin-herng committed:
Test TaskList

chin-herng committed:
Merge branch 'branch-A-JUnit'

Greg-Tan commented on src/main/java/bob/Bob.java:
```diff
@@ -0,0 +1,47 @@
+                command.execute(ui, storage, tasks);
+                isExit = command.isExit();
+            } catch (BobException e) {
+                ui.show(e.getMessage());
```
great usage of try and catch block to handle your errors

chin-herng replied:
sflr thanks!

Greg-Tan commented on src/main/java/bob/Parser.java:
```diff
@@ -0,0 +1,182 @@
+import bob.command.ListCommand;
+import bob.command.ListDueInCommand;
+import bob.command.ListOnDateCommand;
+import bob.command.MarkCommand;
```
should arrange your imports in alphabetical order if possible

chin-herng replied:
sflr, but isn't the highlighted imports already in alphabetical order?

Greg-Tan commented on src/main/java/bob/Parser.java:
```diff
@@ -0,0 +1,182 @@
+    private static String[] extractParameters(String parametersString,
+                                             String[] parameters) throws ParameterNotFoundException {
+        // TODO: generalise the method to any given parametersString, desiredParameters and separator
+        // TODO: perform ParameterNotFoundException checks correctly
```
should remove comments before making the pull request

chin-herng replied:
sflr, I do think that these comments are helpful to communicate the loose ends to tie before the PR can be merged, or possible improvements after the PR is merged, which is why I left it in the PR.

Greg-Tan commented on src/main/java/bob/Bob.java:
```diff
@@ -0,0 +1,47 @@
+        runCommandLoopUntilExitCommand();
+    }
+
+    public static void main(String[] args) {
```
exceptional OOP design to your chatbot!

chin-herng replied:
sflr thanks!

Greg-Tan reviewed (commented):
LGTM. Please remember to remove unnecessary comments before making your pull requests :)

chin-herng committed:
build.gradle: configure mainClassName property

chin-herng committed:
Merge branch 'branch-A-Jar'

chin-herng committed:
Add JavaDoc header comments to non-private classes/methods

chin-herng committed:
Add JavaDoc header comments to all non-trivial private methods

chin-herng committed:
Rename methods to be verbs

chin-herng committed:
Rename test method names

chin-herng committed:
Rename boolean methods to sound like booleans

chin-herng committed:
Rename filteredTask to use plural form

chin-herng committed:
Sort and rename constants to have a common prefix

chin-herng committed:
Add line break to keep line length within 120 chars

chin-herng committed:
Change placement of line break to improve readability

chin-herng committed:
Introduce comment to most logical units

chin-herng committed:
Implement find

chin-herng committed:
List imported classes in bob.command explicitly

chin-herng committed:
Remove list tests

chin-herng committed:
Merge branch 'branch-A-JavaDoc'

chin-herng committed:
Merge branch 'branch-A-CodingStandard'

Resolve merge conflict

chin-herng committed:
Merge branch 'branch-Level-9'

Resolve merge conflicts

chin-herng committed:
Add comments to implementation of find

chin-herng committed:
Remove unnecessary "Parser."

chin-herng committed:
Update build.gradle

chin-herng committed:
Start off GUI programming

chin-herng committed:
Configure Checkstyle

chin-herng committed:
Update build.gradle

chin-herng committed:
Fix coding style violations

chin-herng committed:
Merge branch 'branch-A-CheckStyle'

chin-herng committed:
Add test for find

chin-herng committed:
Merge branch 'master' into branch-Level-10

chin-herng committed:
Produce fully functional GUI prototype

chin-herng committed:
Replace getResponse stub

chin-herng committed:
Remove unusued code

chin-herng committed:
Redefine Ui to be utility class

chin-herng committed:
Add test for getListResponse

chin-herng committed:
Replace AnchorPane with VBox

chin-herng committed:
Replace xmlns value of MainWindow to use version 17

chin-herng committed:
DialogBox.fxml: Add padding between ImageView and Label

chin-herng committed:
DialogBox.java: Clip ImageView into a circle

chin-herng committed:
Add background color to each dialog box

chin-herng committed:
Customise appearance with CSS

chin-herng committed:
Set title and disable resize for stage

chin-herng committed:
Merge branch 'branch-Level-10'

chin-herng committed:
Update README.md

chin-herng committed:
Remove outdated comment

chin-herng committed:
Parser.java: rename parameter command to input

chin-herng committed:
Parser.java: move check for EXIT_COMMAND into switch-case

chin-herng committed:
Parser.java: avoid recycling parameter in parse()

chin-herng committed:
Add @return for JavaDoc header of execute

chin-herng committed:
Enable assertions for Gradle

chin-herng committed:
Use assertions

chin-herng committed:
Parser.java: extract parseListOn and parseListDueIn

parseList is long.

Long methods hinders code readability.

Let's pull up the parsing for "list /on" and "list /due_in" into two
separate methods named parseListOn and parseListDueIn respectively.

Since ListOnCommand and ListDueInCommand inherits from ListCommand,
such extractions make intuitive sense because parseList will call
parseListOn and parseListDueIn to parse more specific types of list
commands.

chin-herng committed:
Parser.java: extract parseAddDeadline and parseAddEvent

parseAdd is long.

Long methods hinders code readability.

Let's pull up the parsing for "deadline" and "event" into two separate
methods named parseAddDeadline and parseAddEvent respectively.

Since AddDeadlineCommand and AddEventCommand inherits from AddCommand,
such extractions make intuitive sense because parseAdd will call
parseAddDeadline and parseAddEvent to parse more specific types of add
commands.

chin-herng committed:
Parser.java: extract extractParameter

extractParameters contains deeply-nested code.

It is difficult to keep track of the logic of deeply-nested code.

Let's pull up the extraction of each parameter into a new method named
parseParameter.

The for loop body in extractParameters will only contain a call to
extractParameter, which reduces the number of levels of indentation.

chin-herng committed:
Use assertions to verify preconditions for various parsing methods

chin-herng committed:
Parser.java: refactor parseDeleteOrMark

chin-herng committed:
Storage.java: extract loadData

chin-herng committed:
Storage.java: extract writeFile

chin-herng committed:
Storage.java: extract rewriteFile

chin-herng committed:
TaskList.java: refactor listWithFilter

chin-herng committed:
Use assertion to verify control-flow invariant for parseAdd

chin-herng committed:
Replace extractParameter with splitByParameter

chin-herng committed:
Fix dialog label height being changed by scroll pane

chin-herng committed:
Ui.java: refactor getAddResponse and getDeleteResponse

chin-herng committed:
DialogBox.java: refactor constructor

chin-herng committed:
Refactor MainWindow.java

chin-herng committed:
Parser.java: refactor extractParameter

chin-herng committed:
Fix loading error response not displaying and add footer

chin-herng committed:
Miscellaneous refactoring and clean up

chin-herng committed:
Revert refactoring of listWithFilter

chin-herng committed:
Use streams

chin-herng committed:
Merge pull request #3 from chin-herng/branch-A-CodeQuality

Branch a code quality

chin-herng committed:
Merge master and resolve merge conflicts

chin-herng committed:
Merge pull request #2 from chin-herng/branch-A-Assertions

Branch a assertions

chin-herng committed:
Merge branch 'master' into branch-A-Streams

chin-herng committed:
Merge pull request #4 from chin-herng/branch-A-Streams

Use streams

chin-herng committed:
Set up CI

chin-herng committed:
Provide support for tasks to be done within a certain period

chin-herng committed:
Merge pull request #5 from chin-herng/branch-B-DoWithinPeriodTasks

Provide support for tasks to be done within a certain period

chin-herng committed:
Fix Checkstyle errors

chin-herng committed:
Add representative screenshot

chin-herng committed:
Remove isOccurringOn check for deadlines

chin-herng committed:
Add test for periods

chin-herng committed:
Add User Guide

chin-herng committed:
Update tests for periods

chin-herng committed:
DeadlineTest.java: remove unnecessary imports
